src/realm/geospatial.hpp

```diff
@@ -0,0 +1,267 @@
+struct GeoPolygon {
+    GeoPolygon(std::vector<GeoPoint>&& p)
+        : points({std::move(p)})
+    {
+    }
+    GeoPolygon(const std::vector<GeoPoint>& p)
+        : points({p})
+    {
+    }
+    GeoPolygon(std::vector<std::vector<GeoPoint>>&& p)
+        : points(std::move(p))
+    {
+    }
+    GeoPolygon(const std::vector<std::vector<GeoPoint>>& p)
+        : points(p)
+    {
+    }
+    GeoPolygon(const GeoPolygon& other) = default;
+    GeoPolygon(GeoPolygon&& other) = default;
+    GeoPolygon& operator=(const GeoPolygon& other) = default;
+    GeoPolygon& operator=(GeoPolygon&& other) = default;
+    GeoPolygon& operator=(std::initializer_list<std::vector<GeoPoint>>&& p)
+    {
+        points = std::move(p);
+        return *this;
+    }
+
+    bool operator==(const GeoPolygon& other) const
+    {
+        return points == other.points;
+    }
+    std::vector<std::vector<GeoPoint>> points;
+};
+
+struct GeoCenterSphere {
+    double radius_radians = 0.0;
+    GeoPoint center;
+
+    bool operator==(const GeoCenterSphere& other) const
+    {
+        return radius_radians == other.radius_radians && center == other.center;
+    }
+
+    // Equatorial radius of earth.
+    static const double c_radius_meters;
+
+    static GeoCenterSphere from_kms(double km, GeoPoint&& p)
+    {
+        return GeoCenterSphere{km * 1000 / c_radius_meters, p};
+    }
+};
+
+class Geospatial {
+public:
+    enum class Type : uint8_t { Point, Box, Polygon, CenterSphere, Invalid };
+
+    Geospatial()
+        : m_value(mpark::monostate{})
+    {
+    }
+    Geospatial(GeoPoint point)
+        : m_value(point)
+    {
+    }
+    Geospatial(GeoBox box)
+        : m_value(box)
+    {
+    }
+    Geospatial(GeoPolygon polygon)
+        : m_value(polygon)
+    {
+    }
+    Geospatial(GeoCenterSphere centerSphere)
+        : m_value(centerSphere)
+    {
+    }
+
+    Geospatial(const Geospatial&) = default;
+    Geospatial& operator=(const Geospatial&) = default;
+
+    Geospatial(Geospatial&& other) = default;
+    Geospatial& operator=(Geospatial&&) = default;
+
+    static Geospatial from_obj(const Obj& obj, ColKey type_col = {}, ColKey coords_col = {});
+    static Geospatial from_link(const Obj& obj);
+    static bool is_geospatial(const TableRef table, ColKey link_col);
+    void assign_to(Obj& link) const;
+
+    std::string get_type_string() const noexcept;
+    Type get_type() const noexcept;
+
+    template <class T>
+    const T& get() const noexcept;
+
+    bool is_valid() const noexcept
+    {
+        return get_type() != Type::Invalid;
+    }
+
+    bool is_within(const Geospatial& bounds) const noexcept;
+    std::string to_string() const;
+
+    bool operator==(const Geospatial& other) const
+    {
+        return m_value == other.m_value;
+    }
+    bool operator!=(const Geospatial& other) const
+    {
+        return !(*this == other);
+    }
+    bool operator>(const Geospatial&) const = delete;
```

Comment:
Why do you need to delete the operators? Is this done for better compiler error messages?

Reply:
Right, we don't want to allow these kinds of comparisons and by deleting them we make sure that they aren't being used anywhere in the query system or elsewhere.